src/Lecture4.hs

```diff
@@ -257,7 +334,10 @@ the file doesn't have any products.
 -}
 
 calculateStats :: String -> String
-calculateStats = error "TODO"
+calculateStats input = case rows of [] -> "the file doesn't have any products"
+                                    _  -> displayStats . combineRows . fromList $ rows
```

Comment:
`rows` could be an empty list and `fromList` throws an error on empty lists. You handle this already by `case` earlier. But you can pattern match on the list and construct `NonEmpty` directly to avoid using partial functions

```suggestion
                                    x : xs  -> displayStats $ combineRows $ x :| xs
```

Reply:
updated, redundant import is deleted